Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs

```diff
@@ -0,0 +1,368 @@
+                _uiDirty = true;
+                //We are done, return. Insert the beaker and exit!
+                SetAppearance(ReagentGrinderVisualState.BeakerAttached);
+                ClickSound();
+                return true;
+            }
+
+            //Next, see if the user is trying to insert something they want to be ground/juiced.
+            if(!heldEnt!.TryGetComponent(out GrindableComponent? grind) && !heldEnt!.TryGetComponent(out JuiceableComponent? juice))
+            {
+                //Entity did NOT pass the whitelist for grind/juice.
+                //Wouldn't want the clown grinding up the Captain's ID card now would you?
+                //Why am I asking you? You're biased.
+                return false;
+            }
+
+            //Cap the chamber. Don't want someone putting in 500 entities and ejecting them all at once.
+            //Maybe I should have done that for the microwave too?
+            if (_chamber.ContainedEntities.Count >= _storageCap)
+            {
+                return false;
+            }
+            _chamber.Insert(heldEnt);
+            _uiDirty = true;
+
+            return true;
+        }
+
+        /// <summary>
+        /// The wzhzhzh of the grinder.
+        /// </summary>
+        /// <param name="isJuiceIntent">true for wanting to juice, false for wanting to grind.</param>
+        private async void DoWork(IEntity user, bool isJuiceIntent)
+        {
+            //Have power, are  we busy, chamber has anything to grind, a beaker for the grounds to go?
+            if(!Powered || _busy || ChamberEmpty || !HasBeaker)
+            {
+                return;
+            }
+
+            _busy = true;
+
+            var chamberContentsArray = _chamber.ContainedEntities.ToArray();
+            var chamberCount = chamberContentsArray.Length;
+
+            //This block is for grinding behaviour only.
+            if (!isJuiceIntent)
+            {
+                UserInterface?.SendMessage(new ReagentGrinderWorkStartedMessage(isJuiceIntent));
+                _audioSystem.PlayFromEntity("/Audio/Machines/blender.ogg", Owner, AudioParams.Default);
+                //Get each item inside the chamber and get the reagents it contains. Transfer those reagents to the beaker, given we have one in.
+                Owner.SpawnTimer(_workTime, (Action) (() =>
+                {
+                    for (int i = chamberCount - 1; i >= 0; i--)
+                    {
+                        var item = chamberContentsArray[i];
+                        if (item.TryGetComponent<SolutionContainerComponent>(out var solution))
+                        {
+                            if (_heldBeaker!.CurrentVolume + solution.CurrentVolume > _heldBeaker!.MaxVolume) continue;
+                            _heldBeaker!.TryAddSolution(solution.Solution);
+                            solution!.RemoveAllSolution();
+                            _chamber.ContainedEntities[i].Delete();
```

Comment:
I'm not entirely sure this is important but you should probably set _chamber.ContainedEntities[i] to null for cleanness sake (also for the garbage collector?), also do this on line 358 for juicing.
Probably doesn't matter though tbh